Typescript types regression starting in version 2.24.81

<!--
  Please fill out each section below, otherwise, your issue will be closed. This info allows Gatsby maintainers to diagnose (and fix!) your issue as quickly as possible.

  Useful Links:
  - Documentation: https://www.gatsbyjs.org/docs/
  - How to File an Issue: https://www.gatsbyjs.org/contributing/how-to-file-an-issue/

  Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
-->

## Description

This issue is a regression in the Typescript types, caused by [PR 27381](https://github.com/gatsbyjs/gatsby/pull/27381), specifically this commit: https://github.com/gatsbyjs/gatsby/pull/27381/commits/2499676068a91c778181a382491ea6c1cc55782c 

It pulls the type from the `src` folder, which is fine in development, but the `src` folder is not included in the distributed npm package so you just get a file missing error.

### Steps to reproduce

Run `tsc --noEmit` on a project including gatsby from npm.

The workaround for now is to revert gatsby and gatsby-plugin-utils.

```sh
npm i gatsby@2.24.80 gatsby-plugin-utils@0.2.34 -E
```

### Expected result

No type error.

### Actual result

```
node_modules/gatsby/index.d.ts:29:36 - error TS2307: Cannot find module './src/bootstrap/load-plugins/types' or its corresponding type declarations.

29 export { IPluginInfoOptions } from "./src/bootstrap/load-plugins/types"
```

### Environment

  System:
    OS: macOS 10.15.7
    CPU: (16) x64 Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
    Shell: 5.0.11 - /usr/local/bin/bash
  Binaries:
    Node: 12.18.4 - ~/.nvm/versions/node/v12.18.4/bin/node
    npm: 6.14.6 - ~/.nvm/versions/node/v12.18.4/bin/npm
  Languages:
    Python: 2.7.16 - /usr/bin/python
  Browsers:
    Chrome: 86.0.4240.80
    Firefox: 80.0
    Safari: 14.0
  npmPackages:
    gatsby: ^2.24.83 => 2.24.83 
    gatsby-image: ^2.4.21 => 2.4.21 
    gatsby-link: ^2.4.16 => 2.4.16 
    gatsby-plugin-canonical-urls: ^2.3.13 => 2.3.13 
    gatsby-plugin-create-client-paths: ^2.3.13 => 2.3.13 
    gatsby-plugin-emotion: ^4.3.14 => 4.3.14 
    gatsby-plugin-feed: ^2.5.17 => 2.5.17 
    gatsby-plugin-google-analytics: ^2.3.19 => 2.3.19 
    gatsby-plugin-manifest: ^2.4.35 => 2.4.35 
    gatsby-plugin-offline: ^3.2.35 => 3.2.35 
    gatsby-plugin-optimize-svgs: ^1.0.3 => 1.0.4 
    gatsby-plugin-react-helmet: ^3.3.14 => 3.3.14 
    gatsby-plugin-react-svg: ^3.0.0 => 3.0.0 
    gatsby-plugin-remove-trailing-slashes: ^2.3.13 => 2.3.13 
    gatsby-plugin-s3: ^0.3.5 => 0.3.5 
    gatsby-plugin-sharp: ^2.6.43 => 2.6.43 
    gatsby-plugin-sitemap: ^2.4.17 => 2.4.17 
    gatsby-plugin-twitter: ^2.3.12 => 2.3.12 
    gatsby-plugin-typescript: ^2.4.23 => 2.4.23 
    gatsby-plugin-typography: ^2.5.13 => 2.5.13 
    gatsby-react-router-scroll: ^3.0.15 => 3.0.15 
    gatsby-remark-autolink-headers: ^2.3.15 => 2.3.15 
    gatsby-remark-copy-linked-files: ^2.3.19 => 2.3.19 
    gatsby-remark-embed-video: ^3.0.10 => 3.0.10 
    gatsby-remark-embedder: ^3.0.0 => 3.0.0 
    gatsby-remark-images: ^3.3.37 => 3.3.37 
    gatsby-remark-prismjs: ^3.5.16 => 3.5.16 
    gatsby-remark-responsive-iframe: ^2.4.17 => 2.4.17 
    gatsby-remark-smartypants: ^2.3.13 => 2.3.13 
    gatsby-source-filesystem: ^2.3.35 => 2.3.35 
    gatsby-source-graphql: ^2.7.6 => 2.7.6 
    gatsby-transformer-remark: ^2.8.44 => 2.8.44 
    gatsby-transformer-sharp: ^2.5.19 => 2.5.19

--- a/packages/gatsby-plugin-utils/src/types.ts
+++ b/packages/gatsby-plugin-utils/src/types.ts
@@ -1,3 +1,11 @@
+/**
+ * ⚠️ DO NOT EDIT ⚠️
+ * These types are duplicated from packages/gatsby/src/bootstrap/load-plugins/types.ts
+ * If you edit this file, make sure to edit that file too!!!
+ * They are duplicate to avoid a circular dependency between gatsby-plugin-utils <=> gatsby <=> gatsby-plugin-utils
+ * See gatsbyjs/gatsby#27578 and ping @laurieontech or @mxstbr if you have any questions
+ */
+
 export interface ISiteConfig {
   plugins?: Array<PluginRef>
 }

--- a/packages/gatsby-plugin-utils/src/types.ts
+++ b/packages/gatsby-plugin-utils/src/types.ts
@@ -1,0 +1,51 @@
+export interface ISiteConfig {
+  plugins?: Array<PluginRef>
+}
+
+// There are two top-level "Plugin" concepts:
+// 1. IPluginInfo, for processed plugins, and
+// 2. PluginRef, for plugin configs
+
+export interface IPluginInfo {
+  /** Unique ID describing a plugin */
+  id: string
+
+  /** The absolute path to the plugin */
+  resolve: string
+
+  /** The plugin name */
+  name: string
+
+  /** The plugin version (can be content hash) */
+  version: string
+
+  /** Options passed to the plugin */
+  pluginOptions?: IPluginInfoOptions
+}
+
+export interface IPluginInfoOptions {
+  plugins?: Array<IPluginInfo>
+  path?: string
+  [option: string]: unknown
+}
+
+export interface IFlattenedPlugin extends IPluginInfo {
+  skipSSR?: boolean
+  ssrAPIs: Array<string>
+  nodeAPIs: Array<string>
+  browserAPIs: Array<string>
+}
+
+export interface IPluginRefObject {
+  resolve: string
+  options?: IPluginRefOptions
+  parentDir?: string
+}
+
+export type PluginRef = string | IPluginRefObject
+
+export interface IPluginRefOptions {
+  plugins?: Array<PluginRef>
+  path?: string
+  [option: string]: unknown
+}

--- a/packages/gatsby-plugin-utils/src/validate.ts
+++ b/packages/gatsby-plugin-utils/src/validate.ts
@@ -1,6 +1,6 @@
 import { ValidationOptions } from "joi"
 import { ObjectSchema } from "./joi"
-import { IPluginInfoOptions } from "gatsby"
+import { IPluginInfoOptions } from "./types"
 
 const validationOptions: ValidationOptions = {
   // Show all errors at once, rather than only the first one every time

--- a/packages/gatsby/index.d.ts
+++ b/packages/gatsby/index.d.ts
@@ -26,8 +26,6 @@ export {
   withAssetPrefix,
 } from "gatsby-link"
 
-export { IPluginInfoOptions } from "./src/bootstrap/load-plugins/types"
-
 export const useScrollRestoration: (
   key: string
 ) => {

